src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.cs

```diff
@@ -0,0 +1,166 @@
+
+    internal DelegateCacheRewriter(SyntheticBoundNodeFactory factory, int topLevelMethodOrdinal)
+    {
+        Debug.Assert(factory.TopLevelMethod is { });
+
+        _factory = factory;
+        _topLevelMethod = factory.TopLevelMethod;
+        _topLevelMethodOrdinal = topLevelMethodOrdinal;
+    }
+
+    internal static bool CanRewrite(CSharpCompilation compilation, MethodSymbol topLevelMethod, bool inExpressionLambda, BoundConversion boundConversion, MethodSymbol targetMethod)
+        => targetMethod.IsStatic
+        && !boundConversion.IsExtensionMethod
+        && !inExpressionLambda // The tree structure / meaning for expression trees should remain untouched.
+        && topLevelMethod.MethodKind != MethodKind.StaticConstructor // Avoid caching twice if people do it manually.
+        && compilation.IsStaticMethodGroupDelegateCacheEnabled
+        ;
+
+    internal BoundExpression Rewrite(int localFunctionOrdinal, SyntaxNode syntax, BoundExpression receiver, MethodSymbol targetMethod, TypeSymbol delegateType)
+    {
+        Debug.Assert(delegateType.IsDelegateType());
+
+        var oldSyntax = _factory.Syntax;
+        _factory.Syntax = syntax;
+
+        var cacheContainer = GetOrAddCacheContainer(localFunctionOrdinal, delegateType, targetMethod);
+        var cacheField = cacheContainer.GetOrAddCacheField(_factory, delegateType, targetMethod);
+
+        var boundCacheField = _factory.Field(null, cacheField);
+        var boundDelegateCreation = new BoundDelegateCreationExpression(syntax, receiver, targetMethod, isExtensionMethod: false, type: delegateType);
+
+        var rewrittenNode = _factory.Coalesce(boundCacheField, _factory.AssignmentExpression(boundCacheField, boundDelegateCreation));
+
+        _factory.Syntax = oldSyntax;
+
+        return rewrittenNode;
+    }
+
+    private DelegateCacheContainer GetOrAddCacheContainer(int localFunctionOrdinal, TypeSymbol delegateType, MethodSymbol targetMethod)
+    {
+        Debug.Assert(_factory.ModuleBuilderOpt is { });
+        Debug.Assert(_factory.CurrentFunction is { });
+
+        var typeCompilationState = _factory.CompilationState;
+        var generation = _factory.ModuleBuilderOpt.CurrentGenerationOrdinal;
+
+        DelegateCacheContainer? container;
+
+        if (CanUseConcreteCacheContainer(delegateType, targetMethod))
+        {
+            container = typeCompilationState.ConcreteDelegateCacheContainer;
+
+            if (container is { })
+            {
+                return container;
+            }
+
+            container = new(typeCompilationState.Type, generation);
+            typeCompilationState.ConcreteDelegateCacheContainer = container;
+        }
+        else
+        {
+            var containersStack = _genericCacheContainers ??= new();
+
+            while (containersStack.Count > 0 && containersStack.Peek().LocalFunctionOrdinal > localFunctionOrdinal)
```

Comment:
>while (containersStack.Count > 0 && containersStack.Peek().LocalFunctionOrdinal > localFunctionOrdinal) [](http://example.com/codeflow?start=12&length=103)

Managing the stack based on local function ordinal feels rather strange to me. It doesn't really provide accurate equivalence of a parent-child relationship. Especially unintuitive feels popping obsolete entries when we are about to push a new one. I think it would be better to pop an entry for a local function when we done visiting its body. So that the top of the stack is always either corresponds to a parent, or to the current function.

Reply:
Yep, I admit this is pretty ugly. Let me see how to change.